hyperspy/_signals/signal2d.py

```diff
@@ -605,3 +608,433 @@ def add_ramp(self, ramp_x, ramp_y, offset=0):
+    """
+    data_max = ndi.filters.maximum_filter(z, separation)
+    maxima = (z == data_max)
+    data_min = ndi.filters.minimum_filter(z, separation)
+    diff = ((data_max - data_min) > threshold)
+    maxima[diff == 0] = 0
+    labeled, num_objects = ndi.label(maxima)
+    peaks = np.array(
+        ndi.center_of_mass(z, labeled, range(1, num_objects + 1)))
+
+    return clean_peaks(peaks)
+
+
+def find_peaks_max(z, alpha=3., size=10):
+    """
+    Method to locate positive peaks in an image by simple local maximum
+    searching.
+    """
+    # preallocate lots of peak storage
+    k_arr = np.zeros((10000, 2))
+    # copy image
+    image_temp = copy.deepcopy(z)
+    peak_ct = 0
+    # calculate standard deviation of image for thresholding
+    sigma = np.std(z)
+    while True:
+        k = np.argmax(image_temp)
+        j, i = np.unravel_index(k, image_temp.shape)
+        if image_temp[j, i] >= alpha * sigma:
+            k_arr[peak_ct] = [j, i]
+            # masks peaks already identified.
+            x = np.arange(i - size, i + size)
+            y = np.arange(j - size, j + size)
+            xv, yv = np.meshgrid(x, y)
+            # clip to handle peaks near image edge
+            image_temp[yv.clip(0, image_temp.shape[0] - 1),
+                       xv.clip(0, image_temp.shape[1] - 1)] = 0
+            peak_ct += 1
+        else:
+            break
+    # trim array to have shape (number of peaks, 2)
+    peaks = k_arr[:peak_ct]
+    return clean_peaks(peaks)
+
+
+def find_peaks_zaefferer(z, grad_threshold=0.1, window_size=40,
+                         distance_cutoff=50):
+    """
+    Method to locate positive peaks in an image based on gradient
+    thresholding and subsequent refinement within masked regions.
+    Parameters
+    ----------
+    z : ndarray
+        Matrix of image intensities.
+    grad_threshold : float
+        The minimum gradient required to begin a peak search.
+    window_size : int
+        The size of the square window within which a peak search is
+        conducted. If odd, will round down to even.
+    distance_cutoff : float
+        The maximum distance a peak may be from the initial
+        high-gradient point.
+    Returns
+    -------
+    peaks : numpy.ndarray
+        (n_peaks, 2)
+        Peak pixel coordinates.
+    Notes
+    -----
+    Implemented as described in Zaefferer "New developments of computer-aided
+    crystallographic analysis in transmission electron microscopy" J. Ap. Cryst.
+    This version by Ben Martineau (2016)
+    """
+
+    def box(x, y, window_size, x_max, y_max):
+        """Produces a list of coordinates in the box about (x, y)."""
+        a = int(window_size / 2)
+        x_min = max(0, x - a)
+        x_max = min(x_max, x + a)
+        y_min = max(0, y - a)
+        y_max = min(y_max, y + a)
+        return np.array(
+            np.meshgrid(range(x_min, x_max), range(y_min, y_max))).reshape(
+            2,
+            -1).T
```

Comment:
using `itertools.product` in my simple test is around 20 times faster (albeit it returns a list of tuples):
```python
from itertools import product
return [a for a in product(range(y_min, y_max), range(x_min, x_max))]
```

(i swapped X and Y order purely to exactly duplicate the order you had previously - `product` iterates the last iterators first )

I'll have a look further in the code, but maybe returning the iterator itself (or maybe even `yield`-ing) may be an even better approach